### PR TITLE
Post: ES6, ES2015 : declaring variables using const, let and var

### PR DESCRIPTION
**Pull request by MoOx**
Translation of @Nyalab post.
Feel free to merge asap so we can release it today.

### CODE REVIEW

File: content/en/articles/js/es2015/const-let-var.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  - ES6
+  - ES2015
+authors:
+  - Nyalab
+translators:
+  - MoOx
+---
+
+ES6 (known as ES2015) brings you new ways to declare your variables thanks to
+`let` and `const` but also keeps `var` in the language specification.
```
Comment (ben-eb) [inline content/en/articles/js/es2015/const-let-var.md]: comma after `const`?

File: content/en/articles/js/es2015/const-let-var.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+The first thing you need to know, is to forget everything you know about `var`.
+
+# Declarations
+
+## const
+
+`const` allows you to declare a single assignment variable lexically binded.
+Sounds weird right? It means that you can assign a variable that can only
+contains a single value, scoped at the block level.
```
Comment (ben-eb) [inline content/en/articles/js/es2015/const-let-var.md]: *contain*

File: content/en/articles/js/es2015/const-let-var.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+  }
+}
+```
+
+You might think a `let` should be used but here but the declaration is evaluated
+on each iteration, so `const` is more appropriate here.
+
+## let
+
+`let` allows you to do the same as `const` without the single assignation
```
Comment (ben-eb) [inline content/en/articles/js/es2015/const-let-var.md]: *single assignation* => *assignment*

File: content/en/articles/js/es2015/const-let-var.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+constraint. So you can understand here that the use-cases are the same as its
+ancestor, `var`.
+By the way, you might often heard or read: `let` is the new `var`.
```
Comment (ben-eb) [inline content/en/articles/js/es2015/const-let-var.md]: *By the way, you might have been told that `let` is the new `var`.*

File: content/en/articles/js/es2015/const-let-var.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+If you already heard stuff about `const`, be careful: those are not immutable
+variables. Only the assigned reference is immutable, not the value.
+This means that the content of an array or an object declared using `const`
+can evolve but the re-assignation of the reference is impossible.
+
+```js
+function fn() {
+  const foo = "bar"
+  if (true) {
+    const foo // SyntaxError, variable need to be assigned with something
```
Comment (wakooka) [inline content/en/articles/js/es2015/const-let-var.md]: needs to be assigned to something

File: content/en/articles/js/es2015/const-let-var.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+  console.log(foo)
+  // "bar", variable belongs to the scope of its block (the "fn" function)
+  console.log(foo2)
+  // "qux"
+}
+```
+
+You can use `let` in loops, the variable used for the iteration will be scoped
+to the block of the loop, not the (parent) scope of the code that contains the
+loop. No more issue with `i` already defined !
```
Comment (ben-eb) [inline content/en/articles/js/es2015/const-let-var.md]: It's better to end the sentence without a space before the exclamation mark. Maybe this last sentence could read:

*This resolves any issues that you may have had with `i` already defined!*

File: content/en/articles/js/es2015/const-let-var.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+  console.log(j)
+  // j is not defined
+}
+fn2() // 0, 1, 2, 3, 4, 5, 6, 7, 8, 9
+```
+
+Note: the example with `const` used in the `for ... of` loop cannot be
+reproduced here.
+The classic `for` loop is imperative, and the declaration is only made once when
+the loop starts. `const` is not relevant is this case.
```
Comment (ben-eb) [inline content/en/articles/js/es2015/const-let-var.md]: *relevant in this case.*

File: content/en/articles/js/es2015/const-let-var.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+```js
+function fn() {
+  const arr = [1, 2, 3]
+  for (const el of arr) {
+    console.log(el)
+  }
+}
+```
+
+You might think a `let` should be used but here but the declaration is evaluated
```
Comment (wakooka) [inline content/en/articles/js/es2015/const-let-var.md]: Two "but" here

File: content/en/articles/js/es2015/const-let-var.md [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+}
+```
+
+JavaScript engine will read all `var` declarations and virtually move those at
+the start of the scope of your function.
+
+`let` and `const` do not benefit of this hoisting mechanism, which can create
+TDZ (Temporal Dead Zone) issues. Variable declaration is not moved, so you might
+encounter situation where a variable does not exist yet. That's the TDZ
+
```
Comment (ben-eb) [inline content/en/articles/js/es2015/const-let-var.md]: This probably needs rewriting.

`let` & `const` do not benefit from variable hoisting, and therefore they can create TDZ (Temporal Dead Zone) issues; these occur when a variable does not exist *yet*.

File: content/en/articles/js/es2015/const-let-var.md [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+write:
+
+```js
+function fn() {
+  console.log(foo) // "bar"
+  var foo = "bar"
+}
+```
+
+JavaScript engine will read all `var` declarations and virtually move those at
```
Comment (wakooka) [inline content/en/articles/js/es2015/const-let-var.md]: move those to

File: content/en/articles/js/es2015/const-let-var.md [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+the start of the scope of your function.
```
Comment (wakooka) [inline content/en/articles/js/es2015/const-let-var.md]: the start of your function scope

File: content/en/articles/js/es2015/const-let-var.md [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+  let foo = "bar"
+}
+```
+
+# Outro
+
+So let's recap:
+
+- Use `const` (might be relevant 99% of the time)
+- If during you development, you need to re-affect the variable, switch to
```
Comment (ben-eb) [inline content/en/articles/js/es2015/const-let-var.md]: *If during development you need to modify the variable, switch to `let`*

File: content/en/articles/js/es2015/const-let-var.md (hunk 9 above)
Comment (wakooka) [inline content/en/articles/js/es2015/const-let-var.md]: @ben-eb  :+1:

**Comment by Nyalab** [comment]
Thanks @MoOx i planned to do it tonight, but you saved me a lot of time by doing this! I will just add some fixes and mix some feedback from people here :smile:

**Comment by Nyalab** [comment]
Thanks @wakooka @ben-eb for the help :smile:. Is it ok for you now guys?

**Comment by ben-eb** [comment]
:+1:

**Comment by Macxim** [comment]
Looks *very* good to me :shipit:!

**Comment by wakooka** [comment]
Looks good to me too :+1:

**Pull request merged**